shared_model/backend/protobuf/impl/proto_query_response_factory.cpp

```diff
@@ -0,0 +1,212 @@
+
+  iroha::protocol::TransactionsResponse *protocol_specific_response =
+      protocol_query_response.mutable_transactions_response();
+  for (const auto &tx : transactions) {
+    protocol_specific_response->add_transactions()->CopyFrom(
+        std::static_pointer_cast<shared_model::proto::Transaction>(tx)
+            ->getTransport());
+  }
+
+  return std::make_unique<shared_model::proto::TransactionsResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::AssetResponse>
+shared_model::proto::ProtoQueryResponseFactory::createAssetResponse(
+    const shared_model::interface::Asset &asset) {
+  iroha::protocol::QueryResponse protocol_query_response;
+
+  iroha::protocol::AssetResponse *protocol_specific_response =
+      protocol_query_response.mutable_asset_response();
+  protocol_specific_response->mutable_asset()->CopyFrom(
+      static_cast<const shared_model::proto::Asset &>(asset).getTransport());
+
+  return std::make_unique<shared_model::proto::AssetResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::RolesResponse>
+shared_model::proto::ProtoQueryResponseFactory::createRolesResponse(
+    const std::vector<shared_model::interface::types::RoleIdType> &roles) {
+  iroha::protocol::QueryResponse protocol_query_response;
+
+  iroha::protocol::RolesResponse *protocol_specific_response =
+      protocol_query_response.mutable_roles_response();
+  for (const auto &role : roles) {
+    protocol_specific_response->add_roles(role);
+  }
+
+  return std::make_unique<shared_model::proto::RolesResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::RolePermissionsResponse>
+shared_model::proto::ProtoQueryResponseFactory::createRolePermissionsResponse(
+    const shared_model::interface::RolePermissionSet &role_permissions) {
+  iroha::protocol::QueryResponse protocol_query_response;
+
+  iroha::protocol::RolePermissionsResponse *protocol_specific_response =
+      protocol_query_response.mutable_role_permissions_response();
+  for (size_t i = 0; i < role_permissions.size(); ++i) {
+    auto perm = static_cast<interface::permissions::Role>(i);
+    if (role_permissions.test(perm)) {
+      protocol_specific_response->add_permissions(
+          shared_model::proto::permissions::toTransport(perm));
+    }
+  }
+
+  return std::make_unique<shared_model::proto::RolePermissionsResponse>(
+      std::move(protocol_query_response));
+}
+
+std::unique_ptr<shared_model::interface::BlockQueryResponse>
+shared_model::proto::ProtoQueryResponseFactory::createBlockQueryResponse(
+    const shared_model::interface::Block &block) {
+  iroha::protocol::BlockQueryResponse protocol_query_response;
+
+  iroha::protocol::BlockResponse *protocol_specific_response =
+      protocol_query_response.mutable_block_response();
+  const auto &proto_block =
+      static_cast<const shared_model::proto::Block &>(block);
```

Comment:
Maybe explicitly set field without dependency on backend proto?

Reply:
Well, yes there's no significant problem here. Although the code with casting and getting internal transport look not that good